### PR TITLE
chore: integrated event bus datapipeline

### PR DESCRIPTION
**Pull request by Shahroz16**
closes: [MBL-345 : Integrate Event Bus in Data Pipelines](https://linear.app/customerio/issue/MBL-345/integrate-event-bus-in-data-pipelines)

integrated `EventBus` with DataPipeline module and added tests.

### CODE REVIEW

File: core/src/main/kotlin/io/customer/sdk/communication/EventBus.kt [hunk 1]
```diff
@@ -32,8 +39,7 @@ class EventBusImpl(
 
     val jobs = mutableListOf<Job>()
 
-    val scope: CoroutineScope = SDKComponent.androidSDKComponent?.scopeProvider?.eventBusScope
-        ?: CoroutineScope(Dispatchers.Default + SupervisorJob())
+    val scope: CoroutineScope = SDKComponent.androidSDKComponent?.scopeProvider?.eventBusScope ?: CoroutineScope(Dispatchers.Default + SupervisorJob())
```
Comment (mrehan27) [inline core/src/main/kotlin/io/customer/sdk/communication/EventBus.kt:42]: I think there should be consistent way to get scope here. As with this code, if the EventBus is somehow accessed before SDK is initialized, the scope would be different.

I just checked, and I think we can safely move the scope to `SDKComponent` for guaranteed access. Is there a reason for restricting it to `AndroidSDKComponent` that I missed? 🤔

File: core/src/main/kotlin/io/customer/sdk/communication/EventBus.kt [hunk 2]
```diff
@@ -55,4 +61,17 @@ class EventBusImpl(
         jobs.forEach { it.cancel() }
         jobs.clear()
     }
+
+    override fun <T : Event> subscribe(type: KClass<T>, action: suspend (T) -> Unit): Job {
+        val job = scope.launch {
+            flow.filter { type.isInstance(it) }.mapNotNull { it.safeCast(type) }.collect { event ->
```
Comment (mrehan27) [inline core/src/main/kotlin/io/customer/sdk/communication/EventBus.kt:67]: Feels like unnecessary looping. Can we do something like this?

```suggestion
            flow.mapNotNull { it.safeCast(type) }.collect { event ->
```

File: core/src/main/kotlin/io/customer/sdk/communication/EventBus.kt [hunk 3]
```diff
@@ -55,4 +61,17 @@ class EventBusImpl(
...
+                action(event)
+            }
+        }
+        jobs.add(job)
+        return job
+    }
+    private fun <T : Event> Event.safeCast(type: KClass<T>): T? {
+        return if (type.isInstance(this)) type.cast(this) else null
```
Comment (mrehan27) [inline core/src/main/kotlin/io/customer/sdk/communication/EventBus.kt:75]: Can't we use Kotlin's predefined safeCast here like this?

```suggestion
        return type.safeCast(this)
```

Or may be remove this method and just use Kotlin one 🤔 

```kotlin
IdentifyEvent::class.safeCast(event)
```

**Review by mrehan27 — APPROVED** [review]
Mostly looks good, just some comments that we should address before merging.

File: core/src/main/kotlin/io/customer/sdk/communication/EventBus.kt (hunk 1 above)
Comment (Shahroz16) [reply core/src/main/kotlin/io/customer/sdk/communication/EventBus.kt:42]: Makes sense, I wanted to keep it in `AndroidSDKComponent` so the modules who don't need it won't initialize then `AndroidSDKComponent` and we won't have this dependency added to them for no reason. But scope and dispatchers seems more of reusable component so happy to move there.

File: core/src/main/kotlin/io/customer/sdk/communication/EventBus.kt (hunk 2 above)
Comment (Shahroz16) [reply core/src/main/kotlin/io/customer/sdk/communication/EventBus.kt]: good catch, it was left from when I was force casting 😄

File: core/src/main/kotlin/io/customer/sdk/communication/EventBus.kt (hunk 3 above)
Comment (Shahroz16) [reply core/src/main/kotlin/io/customer/sdk/communication/EventBus.kt]: I used it first but then I saw this
```
@WasExperimental(ExperimentalStdlibApi::class)
```
and thought it said experimental but it said `was` :D

**Pull request merged**